include/libtorrent/packet_pool.hpp

```diff
@@ -84,26 +87,33 @@ namespace libtorrent
 		// deleter for std::unique_ptr
 		void operator()(packet* p) const
 		{
+			TORRENT_ASSERT(p != nullptr);
 			p->~packet();
 			std::free(p);
 		}
 	};
 
-	inline packet* create_packet(int size)
+	inline packet* create_packet(int const size)
 	{
 		packet* p = static_cast<packet*>(std::malloc(sizeof(packet) + size));
+		if (p == nullptr) aux::throw_ex<std::bad_alloc>();
 		new (p) packet();
+		p->allocated = aux::numeric_cast<std::uint16_t>(size);
```

Comment:
@ajax16384 this is a critical part of the patch. ``allocated`` is set to actually say how much memory is available